### ISSUE TITLE
Configure tests to run on Travis

### ISSUE THREAD

**Issue by IwanVosloo** [reahl/reahl#54]
We have to figure our how to configure travis and how to run our tests there.

**Comment by IwanVosloo** [comment]
I in [IwanVosloo/reahl/tree/travis](https://github.com/IwanVosloo/reahl/tree/travis) I have made some progress:
- all the basic stuff we need installed, including chromium-browser and chromedriver 
- gpg key for testing is imported
- ssh key is created and localhost is added to known_hosts
- xvfb started and DISPLAY set up for headless operation

TODO:
- We will have to undo #47 now, as that workaround was not necessary....
- Fix the bootstrap.py script to deal with its two phases without user input as per commented-out code in .travis.yml (I think a simple check for its single arg is good enough - no need for arg parsing stuff)
- ... and getting the rest of the commented-out stuff to work

NOTE: I had to move on to using a trusty travis environment because in a container you can only install apt packages via the apt addon (if I understand correctly), but not all packages we need can be installed via the addon, so...

**Comment by IwanVosloo** [comment]
I have fiddled a lot with this. I think there is only one problem left: the browser based testing.

In fact, in a newly created chroot on my local machine tests are not passing anymore. This has to do with changes in newer versions of chromium-browser or chromedriver or selenium.

I have played with using firefox as well, and phantomjs. (For phantomjs I have added a method to the Fixture, and you need to apt-get install phantomjs. Note that phantomjs is not whitelisted on travis for apt.)

Stuff break with any of these. And, different things break locally than do on Travis. Tests also seem to block for long times on travis. I have not investigated the cause yet.

So, while I don't know all the issues yet, one issue has to do with the fact that we cannot dismiss alerts anymore. This is true for phantomjs, chrome and firefox. See test: reahl/web_dev/bootstrap/test_files.py:prevent_form_submit . I have not fixed this, but the only sensible workaround seems to be to temporarily stud out the javascript alert method. See: http://stackoverflow.com/questions/15708518/how-can-i-handle-an-alert-with-ghostdriver-via-python

Firefox has issues with triggering focus/focusout and onchange events - when it is not the active window. When it is the active window, these work as expected. See, for example the discussion: https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/157

It is recommended that we use native events in firefox to overcome that issue, but locally I am already on FF 48, and I have seen at least one post (that I cannot find anymore) alleging that FF>=48 will not support native events anymore. Try as I might, I could not get native events switched on on my local FF, so this corroborates that post somewhat - if I can believe the ways you check. (See troubleshooting tips right at the bottom of: https://github.com/SeleniumHQ/selenium/wiki/Native-Events-On-Linux )

It might be that native events may still get past this issue on travis, because travis uses an older FF (31) - and we have some control over which version it should use. (See build 50 on travis for my pull request for testing this). However, since it is going away I would not want to waste time on that strategy If we can help it.

Either way - the FF people are moving away from the previous webdriver implementation to marionette: see: http://seleniumsimplified.com/2016/04/how-to-use-the-firefox-marionette-driver/

Marionette seems promising because it will allow you not only to control the HTML elements, but also the browser menus etc. (Maybe popups??) But... its not there yet - not in a way we can easily get going on Travis.

So in summary:
There may be more problems, but I have identified the following ones:
- All browsers now have an issue with alerts - alerts just makes them unresponsive forever (but there's a workaround).
- FF has issues triggering focusout events (eg) when the browser window is not in focus.
- I know FF had issues running our fancy JS concurrency tests (like test_files.py:async_remove) that is why we moved to chromium before. I don't know what these issues are or why they are issues.
- Seems like the newer chromium now has similar issues as above...
- It concerns me that we are testing locally with different versions of browsers than on travis.

Next steps: to keep pushing to travis is becoming a bit much. I think its time to reproduce a travis test run more truthfully on a local machine. Here is a link to doing that: http://stackoverflow.com/questions/21053657/how-to-run-travis-ci-locally

**Comment by IwanVosloo** [comment]
BTW, [build 50 on travis](https://travis-ci.org/reahl/reahl/builds/162408820) did quite well. It ran with FF 31, native events enabled except tests that explicitly use the chrome driver (probably only those fancy file upload tests?).  Of course, if sticking to an older FF works without much effort I am all for it - until the world hops onto marionette. Then we'll have to just bring our local environment in line.

**Comment by IwanVosloo** [comment]
Just a note on what we figured out yesterday. It turned out that tests ran fine with the version of chromium and chromedriver on our local development chroots but that is only because we do not update the chroots that often. On the latest chromium + chromedriver things break.

For now, we have decided to pick a version combo of these that work, and added scripts for us to be able to install these specific versions locally and on travis.

In future it will probably make sense to see what firefox + marionette can do.

**Comment by IwanVosloo** [comment]
I've added py35 to the .travis.yaml and had to make a change or two to get it green. I also expanded the build to actually build the packages after the test run, start up a devpi server and upload the packages to it. I am also trying to run the tests via tox from there in clean virtualenvs. 

This is tricky because travis already runs different environments - there is not really a need for tox per se. But devpi test expects tox.ini. So, I've added a commandline argument to our reahl devpitest command that will invoke tox only for a single python environment - the one we're currently on.

Somehow tests now get run twice, so there's probably something wrong...

**Comment by IwanVosloo** [comment]
So, after fiddling more with this, here's what I think I understand:

Firstly, devpi test installs and run tests for each package it finds on devpi, so it will run tests once for the tar.gz and another time for the wheel. Thus, this is by design.

The more tricky bits has to do with dependencies, how we want to specify them, how pip resolves them and how we bootstrap our environment.

Pip's dependency resolution is an ongoing problem: https://github.com/pypa/pip/issues/988 (so, we must be aware of its limitations and work around...)

Pip installs the first dependency spec it finds, so, for example:
- we have reahl-webdev depending on WebTest which in turn depends on waitress >0.8.5; and
- we have reahl-dev depending on devpi-server which in turn depends on waitress <1,>0.8.9

In this scenario, when installing reahl-webdev, pip will install the first requirement it finds mentioning waitress (>0.8.5) - thus it takes the latest available which is currently 1.0.0. When it hits the second one (<1,>0.8.9) it breaks because 1.0.0 was installed and that conflicts with <1. The only way for us to solve this right now using our setuptools dependencies is to put a spurious dependency on waitress in reahl-webdev/.reahlproject with the correct version that will not conflict with the later sub, sub dependency encountered via reahl-webdev >> reahl-dev >> devpi-server >> waitress.

To figure something like this out, it helps to create a clean virtualenv, bootstrap and then do a reahl-control listdeps etc/ on the project that is erroring. 

The other issue we have is with bootstrapping. We use bootstrap.py locally for all kinds of reasons. We often stay in one virtualenv, checkout a different version and run bootstrap.py to to make it all work again. Since the bootstrap script itself is dependent on reahl-dev (and all its dependencies) we first need to install the script's dependencies. But, we have to install the correct versions that will not conflict with the set of all the rest we need to install later. This means we need to figure out the correct versions to install and we used to do that by reading the generated reahl_dev.egg-info/requires.txt which we also check in. This plan creates problems when we change the versions of things etc. Also, it only reads the dependencies of reahl-dev itself. Yet, we need its dependencies recursively. We don't want to also store YET another egg_info directory in version control, (for each recursive dependency) to make this work. To fix this I have changed the script to NOT rely on what is in requires.txt but rather to parse the .reahlproject file xml of all core projects and install all their 3rdpartyegg dependencies using the version specs listed there.

Since we first install the script dependencies, and then run the script to do the rest of all the dependencies, it means we install stuff twice during bootstrapping. If you do this, the second run can stuff up what you installed during the first run, by, for example upgrading something installed during the first run. For this reason I did two things: (a) I include all the basic dependencies from the first run in the arguments to pip install when I do the second run (Pip respects what you pass on its commandline); and (b) I do NOT specify --upgrade on the second run.

--upgrade is given on the first run, because on travis we want to reset things to the versions we want instead of working with whatever is already installed on travis.

We also never pass --ignore-installed because we WANT things to break if we install conflicting stuff on subsequent pip install runs.

Either way, in all of this: note that the bootstrapped virtual env on travis is only ever used to build the packages uploaded to the devpi server. The tests run on environments created from scratch by tox solely based on the metadata from our .reahlproject files and totally unrelated to our bootstrap script. We run tests via tox specifically because doing so will highlight dependency issues we missed in the bootstrapped environment where everything is installed togetheri. Luckily running tox locally breaks in the same way it does on travis when you hit such dependency issues making it possible to reproduce and thus debug locally.
